### PR TITLE
Surface Arc-Length PE: curvilinear position encoding for surface nodes

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The TE coordinate frame (PR #2207, -5.4% p_in) proved that explicit geometric reference frame features dramatically improve predictions. That feature provides **radial information** (how far from the TE, in x and y). 

The orthogonal question is: **where along the airfoil surface is this node?** Arc-length from the leading edge, normalized by chord, provides the curvilinear surface coordinate `s ∈ [0, 1]` — the canonical parameterization aerodynamicists use for Cp curves. This is NOT inferrable from DSDF features (which encode distance TO the surface, not ALONG it).

**Physical motivation:**
- For NACA6416 OOD: the leading-edge suction peak occurs at a different chordwise position than NACA0012 due to camber. Arc-length PE (s ∈ [0,1] normalized by chord) is **invariant to chord scale differences** between the two airfoil families.
- The DSDF features (channels 2-5) encode shortest distance to the surface but do not encode where on the surface the nearest point is. Arc-length fills this gap.
- Following TE coord frame success: the model now knows "how far from the TE" (radial). Arc-length PE tells it "how far along the surface from the LE" (curvilinear). Together they triangulate surface location precisely.

**Expected improvement:** -1 to -3% p_oodc (OOD Cp distribution), -0.5 to -2% p_tan.

### 1. Add argparse flag

```python
parser.add_argument('--surface_arc_length_pe', action='store_true', default=False,
                    help='Add arc-length from LE (normalized by chord) as input feature for surface nodes')
```

### 2. Implement in `Transolver.forward` (after the wake_deficit_feature block, before `preprocess` GatedMLP2)

```python
if cfg.surface_arc_length_pe:
    # Surface node masks (boundary_id 5/6 = fore-foil, 7 = aft-foil)
    fore_surf = (boundary_id == 5) | (boundary_id == 6)  # [B, N]
    aft_surf = (boundary_id == 7)  # [B, N]

    xy = x[:, :, 0:2]  # [B, N, 2] node coordinates

    def compute_arc_length(surf_mask):
        """Compute normalized arc-length for surface nodes.
        Returns [B, N] with arc-length in [0,1] for surface nodes, 0 for volume nodes."""
        B, N = surf_mask.shape
        arc = torch.zeros(B, N, device=x.device)

        for b in range(B):
            idx = surf_mask[b].nonzero(as_tuple=True)[0]  # surface node indices
            if len(idx) == 0:
                continue
            pts = xy[b, idx, :]  # [K, 2]

            # Sort by angle from centroid (robust LE-to-TE ordering)
            centroid = pts.mean(0)
            angles = torch.atan2(pts[:, 1] - centroid[1], pts[:, 0] - centroid[0])
            order = angles.argsort()
            pts_sorted = pts[order]

            # Cumulative arc-length
            diffs = pts_sorted[1:] - pts_sorted[:-1]
            seg_len = diffs.norm(dim=-1)
            cumlen = torch.cat([torch.zeros(1, device=x.device), seg_len.cumsum(0)])
            total = cumlen[-1].clamp(min=1e-6)
            s = cumlen / total  # [K] in [0, 1]

            # Un-sort and assign
            unsort = order.argsort()
            arc[b, idx] = s[unsort]
        return arc

    arc_fore = compute_arc_length(fore_surf)  # [B, N]
    arc_aft = compute_arc_length(aft_surf)    # [B, N]
    arc_total = arc_fore + arc_aft  # non-overlapping masks, so sum is correct
    x = torch.cat([x, arc_total.unsqueeze(-1)], dim=-1)  # [B, N, D+1]
```

**Implementation note on n_x:** The `n_x` variable controls the input dimension for `preprocess` Linear. Add `n_x += 1` when `--surface_arc_length_pe` is enabled, alongside the existing `te_coord_frame` and `wake_deficit_feature` adjustments.

**Performance note:** The per-sample loop above is correct but slow for large batches. Vectorize if possible, but correctness is more important than speed for the first run. The loop processes ~batch_size=4 samples, each with ~150 surface nodes — overhead should be manageable.

**Alternative faster implementation:** Sort surface nodes by x-coordinate (instead of angle from centroid) as a simpler proxy for LE-to-TE ordering. This works for NACA airfoils where LE is at x_min and TE is at x_max. Upper and lower surfaces can be separated by sign of y:

```python
# Simpler alternative: x-coord based sorting for NACA foils
for b in range(B):
    idx = surf_mask[b].nonzero(as_tuple=True)[0]
    if len(idx) == 0:
        continue
    pts = xy[b, idx, :]
    # Sort by x (LE to TE)
    order = pts[:, 0].argsort()
    pts_sorted = pts[order]
    diffs = pts_sorted[1:] - pts_sorted[:-1]
    seg_len = diffs.norm(dim=-1)
    cumlen = torch.cat([torch.zeros(1, device=x.device), seg_len.cumsum(0)])
    s = cumlen / cumlen[-1].clamp(min=1e-6)
    unsort = order.argsort()
    arc[b, idx] = s[unsort]
```

Use whichever approach produces sensible arc-length values. Check that s values are in [0,1] and vary smoothly along the surface.

### 3. Run commands (2 seeds)

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent fern --seed 42 \
  --wandb_name "fern/surface-arc-pe-s42" --wandb_group "fern/surface-arc-length-pe" \
  --surface_arc_length_pe \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature

# Seed 73: same with --seed 73 --wandb_name "fern/surface-arc-pe-s73"
```

## Baseline

**Current best (PR #2213 — Wake Deficit Feature, 2-seed avg):**

| Metric | Value | Merge threshold |
|--------|-------|-----------------|
| p_in | **11.979** | < 11.98 |
| p_oodc | **7.643** | < 7.65 |
| p_tan | **28.341** | < 28.34 |
| p_re | **6.300** | < 6.30 |

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-wake-deficit" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
## Experiment Results: Surface Arc-Length PE

### Setup
- **Branch:** `fern/surface-arc-length-pe`
- **Flag:** `--surface_arc_length_pe` (+ all baseline flags + TE coord frame + wake deficit)
- **Epochs:** 144-145 (180-min timeout)
- **W&B runs (offline):** `nbi3l00s` (s42), `a2dbpn02` (s73)

### Results — Surface Pressure MAE (lower is better)

| Split | Baseline avg | Arc-PE s42 | Arc-PE s73 | Arc-PE avg | Δ |
|-------|-------------|-----------|-----------|-----------|---|
| p_in | 11.979 | **11.8** | 12.9 | 12.35 | +3.1% ❌ |
| p_tan | 28.341 | 28.4 | **28.3** | 28.35 | ~0% ➖ |
| p_oodc | 7.643 | 8.1 | 7.9 | 8.00 | +4.7% ❌ |
| p_re | 6.300 | 6.6 | 6.7 | 6.65 | +5.6% ❌ |

### Implementation Details
- Computes normalized arc-length s ∈ [0, 1] for each foil's surface nodes separately (fore: saf_norm ≤ 0.005, aft: saf_norm > 0.005)
- Sorts surface nodes by angle from centroid for consistent LE-to-TE ordering
- Cumulative arc-length normalized by total perimeter
- Volume nodes get 0
- Added as +1 input channel to the preprocess MLP

### Analysis

**Surface arc-length PE did NOT beat baseline.** Results are essentially neutral on p_tan but consistently worse on the other three splits.

**Why it likely didn't help:**

1. **Redundancy with existing features.** The model already has TE coordinate frame features (dx, dy, r from each TE) and Fourier positional encoding of (x, y). Together these already encode "where along the surface" — adding a 1D arc-length parameter provides little new information. The TE coord frame gives a 6D rich description of surface position; a 1D scalar adds minimal signal.

2. **Angle-based sorting is approximate.** Sorting surface nodes by angle from centroid gives a reasonable ordering for convex NACA airfoils but is imperfect for cambered foils (NACA6416) where the upper and lower surfaces can overlap in angle space. This introduces noise in the arc-length values rather than a clean monotonic signal.

3. **Single feature channel has limited capacity.** The +1 channel is small compared to the existing 56+ input features. The preprocess MLP may struggle to extract useful signal from it, especially given the noise from approximate ordering.

4. **p_oodc and p_re regression suggests slight optimization interference.** The extra feature channel adds a small number of parameters to the first MLP layer. With the same training budget (~145 epochs), the model may not have had enough time to fully integrate the new feature, leading to slight underfitting on OOD splits.

### Suggested Follow-ups

- **Not recommended to pursue further** in current form — the feature is largely redundant with TE coord frame + Fourier PE.
- If revisited: (1) use x-coordinate sorting instead of angle sorting for a cleaner signal, (2) encode as sin/cos of arc-length (Fourier basis) for richer representation, (3) test WITHOUT te_coord_frame to see if arc-length can substitute.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2223 because ADVISOR: Closing as dead end. p_oodc +4.7%, p_re +5.6%, p_in +3.1%, p_tan flat.

Your analysis nails it — the 1D arc-length scalar is largely redundant with the 6D TE coordinate frame that already precisely triangulates surface position. Additionally, the angle-based sorting introduces noise for cambered foils (NACA6416) where upper/lower surfaces can overlap in angle space.

Good implementation and thorough analysis. The observation that this is mostly redundant with existing features is important — it tells us the model already has good surface position information from TE coord frame + Fourier PE. Moving you to a training procedure experiment..
